Fix handshaker API to safely handle shutdown

markdroth opened the pull request:
This is the first part of addressing #8744.  It includes:

- Clarifies semantics of endpoint hand-off from handshaker API.
- Schedule handshaker callbacks via exec_ctx instead of invoking directly.
- Move deadline handling into handshake manager, so individual handshakers don't need to worry about it.

A subsequent PR will change the security handshakers to use the generic handshaker API.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/grpc/grpc/8782)
<!-- Reviewable:end -->

markdroth commented:
Filed issue #8904 to track the endpoint problem.

jboeuf commented on src/core/ext/client_channel/http_connect_handshaker.c:
```diff
@@ -83,65 +94,108 @@ static void http_connect_handshaker_unref(http_connect_handshaker* handshaker) {
     // Otherwise, read the response.
-    grpc_endpoint_read(exec_ctx, handshaker->endpoint, handshaker->read_buffer,
+    // The read callback inherits our ref to the handshaker.
+    grpc_endpoint_read(exec_ctx, handshaker->args->endpoint,
```
Just making sure here that we are confident that `on_read_done` will not be called synchronously which would result in a deadlock.

yang-g commented:
Reviewed 6 of 8 files at r1, 1 of 2 files at r2.
Review status: 6 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 120 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrN0Q9IEhLntrYEKhq:-KXrN0QAItiWAv1f52eD:btjl9mw) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L120)):*
> ```C
>     error = GRPC_ERROR_CREATE("Handshaker shutdown");
>   }
>   if (!handshaker->shutdown) {
> ```

Do we want to set handshaker->shutdown = true here? Otherwise we can shutdown the endpoint twice?

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 231 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrMmDISlplW2-0cRG6:-KXrMmDISlplW2-0cRG7:b-wzrul5) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L231)):*
> ```C
>   grpc_exec_ctx_sched(exec_ctx, handshaker->on_handshake_done, error, NULL);
> done:
>   gpr_mu_unlock(&handshaker->mu);
> ```

What happens if at this unlock, the manager's timer fired and calls connect handshaker's shutdown?

---

*[src/core/lib/channel/handshaker.h, line 136 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrA6Nn6vI0CQ17yINl:-KXrA6Nn6vI0CQ17yINm:bkt3fp) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/lib/channel/handshaker.h#L136)):*
> ```objc
> /// Takes ownership of \a endpoint, and then passes that ownership to
> /// the \a on_handshake_done callback.
> /// Does NOT take ownership of \a args.  Instead, makes a copy before
> ```

s/args/channel_args/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/8782)*
<!-- Sent from Reviewable.io -->

markdroth commented:
Review status: 6 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 120 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrN0Q9IEhLntrYEKhq:-KXrNjbKccS6P0GZRhgQ:b2f6d8m) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L120)):*
<details><summary><i>Previously, yang-g (Yang Gao) wrote…</i></summary><blockquote>

Do we want to set handshaker->shutdown = true here? Otherwise we can shutdown the endpoint twice?
</blockquote></details>

No, handshaker->shutdown should only be set by http_connect_handshaker_shutdown().  The code here is just checking why we're failing.

The handshake_failed_locked() function is called when the read or write callback is invoked either with error != GRPC_ERROR_NONE or with handshaker->shutdown true.  There are actually three cases here:

1. There was no shutdown, but the read or write failed for an unrelated reason.  In this case, error != GRPC_ERROR_NONE and handshaker->shutdown is false.

2. There was a shutdown that interrupted the endpoint read or write.  In this case, error != GRPC_ERROR_NONE and handshaker->shutdown is true.

3. There was a shutdown that occurred after the read or write succeeded but before the read or write callback was invoked.  In this case, error == GRPC_ERROR_NONE but handshaker->shutdown is true.

If the handshaker was already shut down (cases 2 and 3), then the code in http_connect_handshaker_shutdown() has already shut down the endpoint and called cleanup_args_for_failure_locked().  So it's only if the handshaker was NOT already shut down (case 1) that we need to take those steps here.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 231 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrMmDISlplW2-0cRG6:-KXrTSva3sRX3nmLpCKj:b-uczjak) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L231)):*
<details><summary><i>Previously, yang-g (Yang Gao) wrote…</i></summary><blockquote>

What happens if at this unlock, the manager's timer fired and calls connect handshaker's shutdown?
</blockquote></details>

Good catch.  I've changed this to set handshaker->shutdown = true here, so that a subsequent attempt to shut down the handshaker will do nothing.

This also exposed a problem when retrying connections, where the handshakers were being restarted when they might not yet have finished cleaning up the old connection.  I fixed this by changing the connector code to create a new handshake manager for each connection attempt.

---

*[src/core/lib/channel/handshaker.h, line 136 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrA6Nn6vI0CQ17yINl:-KXrSD0j7DpEVxj0eiLU:b-896fix) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/lib/channel/handshaker.h#L136)):*
<details><summary><i>Previously, yang-g (Yang Gao) wrote…</i></summary><blockquote>

s/args/channel_args/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/8782)*
<!-- Sent from Reviewable.io -->

yang-g commented:
Review status: 3 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 120 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrN0Q9IEhLntrYEKhq:-KY-e3NHutg3o0Tadisa:b-rlormb) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L120)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

No, handshaker->shutdown should only be set by http_connect_handshaker_shutdown().  The code here is just checking why we're failing.

The handshake_failed_locked() function is called when the read or write callback is invoked either with error != GRPC_ERROR_NONE or with handshaker->shutdown true.  There are actually three cases here:

1. There was no shutdown, but the read or write failed for an unrelated reason.  In this case, error != GRPC_ERROR_NONE and handshaker->shutdown is false.

2. There was a shutdown that interrupted the endpoint read or write.  In this case, error != GRPC_ERROR_NONE and handshaker->shutdown is true.

3. There was a shutdown that occurred after the read or write succeeded but before the read or write callback was invoked.  In this case, error == GRPC_ERROR_NONE but handshaker->shutdown is true.

If the handshaker was already shut down (cases 2 and 3), then the code in http_connect_handshaker_shutdown() has already shut down the endpoint and called cleanup_args_for_failure_locked().  So it's only if the handshaker was NOT already shut down (case 1) that we need to take those steps here.
</blockquote></details>

I missed the point that once this is called, the handshaker_args->endpoint will become NULL. Does this mean if there is a handshaker shutdown call from upper level after this is called (for reason 1). We will call grpc_endpoint_shutdown with a NULL endpoint? Is that supported?

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 231 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrMmDISlplW2-0cRG6:-KY-itINXm3eSbAqxSQi:b-ijcx02) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L231)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

Good catch.  I've changed this to set handshaker->shutdown = true here, so that a subsequent attempt to shut down the handshaker will do nothing.

This also exposed a problem when retrying connections, where the handshakers were being restarted when they might not yet have finished cleaning up the old connection.  I fixed this by changing the connector code to create a new handshake manager for each connection attempt.
</blockquote></details>

I did not see a shutdown = true but I think you meant handshake_failed_locked? However, does this mean the shutdown will be a no-op if it fires at some specific time point? 

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 251 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KY-ij4nS8wl1wkrmyr0:-KY-ij4nS8wl1wkrmyr1:b36sgri) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L251)):*
> ```C
>   if (!handshaker->shutdown) {
>     handshaker->shutdown = true;
>     grpc_endpoint_shutdown(exec_ctx, handshaker->args->endpoint);
> ```

Maybe add a if (handshaker->args->endpoint != NULL) here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/8782)*
<!-- Sent from Reviewable.io -->

markdroth commented:
Review status: 3 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 120 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrN0Q9IEhLntrYEKhq:-KY-svMazleA5iTa905g:bbskk8a) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L120)):*
<details><summary><i>Previously, yang-g (Yang Gao) wrote…</i></summary><blockquote>

I missed the point that once this is called, the handshaker_args->endpoint will become NULL. Does this mean if there is a handshaker shutdown call from upper level after this is called (for reason 1). We will call grpc_endpoint_shutdown with a NULL endpoint? Is that supported?
</blockquote></details>

Ah, good point.  So yes, we should set shutdown=true here.

Done.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 231 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrMmDISlplW2-0cRG6:-KY-tXW0421y-Ns_Ag8A:bhapho2) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L231)):*
<details><summary><i>Previously, yang-g (Yang Gao) wrote…</i></summary><blockquote>

I did not see a shutdown = true but I think you meant handshake_failed_locked? However, does this mean the shutdown will be a no-op if it fires at some specific time point? 
</blockquote></details>

I'm now setting shutdown=true both here and in handshake_failed_locked.  The idea here is that once we finish the handshake (regardless of whether it was a success or failure), a subsequent shutdown request should be a no-op.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 251 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KY-ij4nS8wl1wkrmyr0:-KY-tmHEoKuJ09Ulm7FM:b-aphq39) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L251)):*
<details><summary><i>Previously, yang-g (Yang Gao) wrote…</i></summary><blockquote>

Maybe add a if (handshaker->args->endpoint != NULL) here?
</blockquote></details>

I don't think this is needed, now that we're setting shutdown to true at both possible exit points from the handshaker.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/8782)*
<!-- Sent from Reviewable.io -->

yang-g commented:
Review status: 3 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 231 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrMmDISlplW2-0cRG6:-KY09Lwf3RxajEinWgga:b-plkj53) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L231)):*
<details><summary><i>Previously, markdroth (Mark D. Roth) wrote…</i></summary><blockquote>

I'm now setting shutdown=true both here and in handshake_failed_locked.  The idea here is that once we finish the handshake (regardless of whether it was a success or failure), a subsequent shutdown request should be a no-op.
</blockquote></details>

If we are finished with this handshaker and queue a on_handshake_done, which will turn to next handshaker. Will a manager shutdown call happening in the middle be ignored because the old handshaker treats it as a no-op and the new handshaker does not know about it?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/8782)*
<!-- Sent from Reviewable.io -->

markdroth commented:
Review status: 3 of 9 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[src/core/ext/client_channel/http_connect_handshaker.c, line 231 at r2](https://reviewable.io:443/reviews/grpc/grpc/8782#-KXrMmDISlplW2-0cRG6:-KY0OPo_85n-P_WhaMD-:b-uk3x1s) ([raw file](https://github.com/grpc/grpc/blob/b8f97a4ac02536bcb88867061b9dad6c35818f2a/src/core/ext/client_channel/http_connect_handshaker.c#L231)):*
<details><summary><i>Previously, yang-g (Yang Gao) wrote…</i></summary><blockquote>

If we are finished with this handshaker and queue a on_handshake_done, which will turn to next handshaker. Will a manager shutdown call happening in the middle be ignored because the old handshaker treats it as a no-op and the new handshaker does not know about it?
</blockquote></details>

Good catch.  I've fixed the handshake manager code to catch this race condition.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/grpc/grpc/8782)*
<!-- Sent from Reviewable.io -->

yang-g commented:
LGTM
